### ISSUE TITLE
Addition of `get_eligible_loading_strategy()` method to surface correct strategy for a given script handle

### ISSUE THREAD

**Issue by 10upsimon** [10up/wordpress-develop#13]
## Description
With the finalisation of the `wp_get_normalized_script_args()` method implementation #10 , including the the script registration process having been updated (see issues [9](https://github.com/10up/wordpress-develop/issues/9) and [8](https://github.com/10up/wordpress-develop/issues/8)), the `get_eligible_loading_strategy()` method can be introduced to the `WP_Scripts` class/API. The current `WP_Scripts` API resides in source [at this location](https://github.com/WordPress/WordPress/blob/6.1-branch/wp-includes/class-wp-scripts.php).

This loading strategy will be determined based on a number of factors, such as whether there are dependents or dependencies, and what their intended strategies are. Checks should be performed against all script dependencies to ensure that they too have loading strategies, before actioning the applicable strategy on said script handle. In essence, this should function is a recursive one that must avoid loops.

While this logic will ultimately first get the *intended* loading strategy, i.e the preferred strategy assigned during script registration, it should call any helper methods required to obtains the least strict loading strategy. 

Strategies are ranked from most to least strict as follows:

- async
- defer
- blocking (default)

Async is considered to be the most strict in the sense that it has the most strict requirements that need to be met in order for the strategy to be applied. Blocking, on the other hand, is considered to be the least strict strategy in the sense that if any of a script's dependencies are blocking, it and all its dependencies become blocking as well, and vice versa.

It should also be noted that WordPress core does not attempt to solve for circular dependencies, and this function is not expected to solve for this either.

## Summary of Loading Strategy Rules
* A script has no dependencies or dependents: `async`
* A script has only blocking or deferred dependencies and only deferred dependents without any non standalone “before” position inline scripts: `defer`
* All other scenarios: `blocking` (although not an actual loading attribute, merely the default behaviour)

## Acceptance Criteria
Create new `get_eligible_loading_strategy()` method within the existing [WP_Scripts](https://github.com/WordPress/WordPress/blob/6.1-branch/wp-includes/class-wp-scripts.php) class.
* Method to be private as only utilize during script registration process (or prior to printing) in order to gauge most eligible loading strategy.
* Method to accept one parameter, `$handle` (string), which is the string identifier of an already registered script handle. Defaults to `''`.
* Introduction of business logic into the recently added `get_eligible_loading_strategy()` method within the existing [WP_Scripts](https://github.com/WordPress/WordPress/blob/6.1-branch/wp-includes/class-wp-scripts.php) class.
* Method should act against the script handle as passed to its $handle parameter,
* Method should check if passed `$handle` exists as a registered script handle, and return a blocking strategy if not, as that is the default.
* Method to utilize the intended loading strategy passed via the applicable script handles `$args` parameter during the use of either `wp_register_script()` or `wp_enqueue_script()` functions, coupled with necessary business logic that checks the dependency tree and adjust the strategy accordingly.
* Method is to be recursive in nature and check all dependents and dependency scripts of a given handle, but should be written in a way that it never enters into an infinite loop.
* Method is to return the most eligible loading and employ the following business rules:
	*  If the script is eligible for an async strategy, the function will apply that since it has the strictest requirements. 
	*  When the intended loading strategy is set to `defer` and the script is eligible for an `async` loading strategy, it should use a  `defer` loading strategy, since if `async` can be used, `defer `certainly can too.
	*  The second potential eligible strategy of a handle is `defer`. If a script is eligible for a `defer` loading strategy, it can use a `blocking` or `defer` loading strategy, but not an `async` loading strategy. When the intended loading strategy for a given script handle is `async` and the script is eligible for a `defer` loading strategy, it should use the `defer` loading strategy. This could be any scenario where an `async` script is depended on by a `defer` script, making the `async` loading strategy on the dependency infeasible.

### Unit Tests
* Unit tests for the various processes script and script dependency logic currently reside in WordPress core source  [at this location](https://github.com/WordPress/wordpress-develop/blob/trunk/tests/phpunit/tests/dependencies/scripts.php).
* Applicable tests should be added (or existing tests modified) that assert that scripts eligible for an `async` strategy contain the `async` loading strategy attribute on the printed `<script>` tag.
* Applicable tests should be added (or existing tests modified) that assert that scripts eligible for a `defer` strategy contain the `defer` loading strategy attribute on the printed `<script>` tag.
* Applicable tests should be added (or existing tests modified) that assert that scripts with a blocking (default) strategy contain no loading strategy attribute on the printed `<script>` tag.
* Applicable tests should be added to ensure that if `CONCATENATE_SCRIPTS` is set to `true`, that loading strategy logic be bypassed altogether, allowing the WP_Scripts API to function exactly as it did prior to the loading strategy work, as if no loading strategy API was present.
* Test(s) should be introduced to ensure that - when checking dependencies for eligible strategies based on the dependency tree - that a recursive loop is never entered into. This needs to be done due to he fact that multiple dependencies across various scripts may depend on one another, and when checking for eligible strategies there is a chance that a recursive loop be entered into.
	*  Tests should track the total amount of script handles registered, and the number of dependency strategy eligibility  checks should never, in theory, exceed the total amount of registered script handles. 
	* Tests to DOT have to solve for actual circular decencies of scripts themselves, as core currently does not handle this scenario at all, but they should test to ensure that circular dependencies are not introduced as a result of work done in this issue

### Code of Conduct

- [X] I agree to follow this project's Code of Conduct

**Comment by joemcgill** [comment]
Sorry, responded to #12 before reading this issue. I'm not sure I understand the value in separating the creation of the method from implementing the business logic unless there is a way to split the business logic implementation into multiple parts, where the initial work would include implementing the least complex use case and this is used to enhance the base implementation with the rest of the logic necessary for calculating the most eligible strategy.

In terms of business rules, the first two bullets seem to be contradictory as written since the first says to use async if eligible, but the second says that defer should be used even if the script is eligible for async. Perhaps this is just my lack of understanding though, so maybe we can specify what is meant by "if eligible" in terms that could be used to write the necessary unit tests that confirm this is implemented as expected?

**Comment by 10upsimon** [comment]
@joemcgill I've combined the #12 into this issue, so I think we can remove #12 now. I left it in the Definition column in case you needed to reference it.

To answer your questions RE points 1 and 2 seeming contradictory. I don't think they are, as there is a difference between the _eligible_ loading strategy and the _intended_ loading strategy. 

What we are saying here is that if the script is _eligible_ for an `async` strategy (based on its dependencies and their dependencies) and does not, then we use `async` as it is the least strict. In this case, I'd imagine all of its dependencies are `async` too. @felixarntz correct me if I am wrong here. However, and this is where point 2 comes into play, when the _intended_ loading strategy is set to `defer` (i.e the user specified that that is the desired loading strategy during registration) BUT it is eligible for an `async` strategy (again, based on logical checks against dependencies) we are to rather use `defer`. This is only when the intention was to `defer`. @felixarntz if I am not making sense here, perhaps you can outline this in a better context than I :)

**Comment by joemcgill** [comment]
Thanks for combining this with #12 and for clarifying the language of business rules. It would be good to include some unit tests at the class level that confirm this is implemented correctly, but those tests can target the public APIs that make use of this underlying method, rather than specifically covering this method itself. Does that make sense?

**Comment by 10upsimon** [comment]
@joemcgill I am not sure I understand the ask, are you stating that I should - in this case.- suggest tests be added to the WP_Scripts or WP_Dependencies class checks to ensure that the correct strategy has been physically applied to a script handle for example?

**Comment by felixarntz** [comment]
@10upsimon @joemcgill Overall the ACs here look solid to me.

+1 to @joemcgill's point on making sure this has solid test coverage (in general, all of this effort needs to be well tested particularly with all the edge cases and backward compatibility concerns).

**Comment by joemcgill** [comment]
Essentially, what I'm getting at is that we'll want unit tests that cover the requirements that you've laid out for this method. However, since this is a private method and not part of the public API, those tests will likely need to be implemented on the public methods of the API that rely on the private method.

That could be accomplished by adding tests for the `WP_Scripts::do_item` method that confirm the behavior of this new `WP_Scripts::get_eligible_loading_strategy` method, as I've suggested in https://github.com/10up/wordpress-develop/issues/14#issuecomment-1428520218, or there may be other ways that make more sense (like calling some method that invokes this private method and then inspecting the properties of affected the `WP_Script` object to confirm the correct loading strategy has been applied).

**Comment by joemcgill** [comment]
@10upsimon, thanks for clarifying the testing criteria here. I think this is good to move to implementation 👍🏻

**Comment by kt-12** [comment]
Making a note of discoveries made while developing this (PR #35), there is also a discussion point for us to brainstorm (if not already done).

I have based my finding on the flow knowledge obtained from ref:https://javascript.info/script-async-defer and the lab test I did to validate it.

### A few changes to the AC - 
**UPDATE: I now see that below mentioned issues are addressed in detail in Milestone 3 (so we can skip this here)** Thanks @joemcgill  for pointing that out.
<strike>
Under `Summary of Loading Strategy Rules`, pointer 2 says _'... without any non standalone “before” position inline scripts: defer'_.  There are a few corrections, as mentioned below -
- `before` inline scripts are always discovered before the discovery of the main script. So this gets executed as intended irrespective of the strategy used by the main script. 
- The main issue is with the usage of `after`. For `async`, the main script and DOM are loaded in parallel; in many cases, `after` will run before the main script. In the case of `defer`, the main script is called when the DOM is ready, so `after` will always be executed before the main script. 
- If a main script has dependent `after` scripts, we should always load it in a blocking way.
 </strike> 


### **Final Rules:**
**Blocking if:**

1. a script has explicit 'strategy => blocking' during registration. 
2. <strike> a script has any 'after' scripts.</strike> (update: will be handled properly in Milestone 3)
3. handle is not present in the registered list. 

**Async if:**

1. There are no dependents or dependencies.
2. The intended strategy was not set during registration or set to `async` 

**Defer if:**

1. All dependents are deferable ( recursion used here)
2. None of the dependencies should be `async` ( however, no dependency having any dependent will be assigned `async` - see rules for `async` above. Indirectly, we say for sure that every dependency has a `blocking` or `defer` strategy. So an explicit check was avoided while coding )

In all other cases, the rule is `blocking`. 

### Now, the issue to think about (if not already thought of):  
@10upsimon  @joemcgill  @felixarntz  I need your thought here

The current code respects the core principle of WPP i.e; we set the best rule applicable by default with a provision to disable it.

Here the question to think about is - 
**What should be the script's strategy if no explicit intended `strategy` is mentioned during registration?**
_**If we go with the best possible strategy:**_
 - Then in many cases, it is going to be `defer`. 
 - The issue is `defer` happens towards the end. So any website dependent on javascript or written in a way that assumes the script will be blocking; will take a hit on the CLS.
 - Even for `async` above is applicable in some very rare cases. 
  
 _**If we go with the default strategy as `blocking` approach**_
- We will only see results as people start using it. Which will be slow, and people can miss out on the apparent optimizations.

**Comment by joemcgill** [comment]
This is really great feedback, @kt-12 and a very important thing for us to document about the behavior of inline scripts set to `after`, though I think that the intention is to handle properly loading inline scripts attached to async/defer scripts in #18, so these changes may not be necessary, unless I'm misunderstanding.

As for your second question, I strongly suspect that we'll need to make this fully opt-in and keep the default as blocking for now, though I could see this as a good thing to think about adding a filter for so that sites could change this default strategy in one place, rather than needing to modify every single script that gets loaded on their site.

**Comment by kt-12** [comment]
Adding a note on rules in the development - PR #35

### **Final Rules:**
**Blocking if:**

1. a script has explicit 'strategy => blocking' or no strategy during registration. 
2. handle is not present in the registered list. 

**Async if:**

1. There are no dependents or dependencies.
2. The intended strategy was not set during registration or set to `async` 

**Defer if:**

1. None of the dependencies should be `async` ( however, no dependency having any dependent will be assigned `async` - see rules for `async` above. Indirectly, we say for sure that every dependency has a `blocking` or `defer` strategy. So an explicit check was avoided while coding )
2. Only deferred dependents <strike> without any non standalone “before” position inline scripts </strike> ( standalone is defined in Milestone 3 issue [#15](https://github.com/10up/wordpress-develop/issues/15) ). Position check will be handled in Milestone 3. 

@10upsimon  in defer we can also consider 'async' dependent along with `defer`, reason being, any `async` script having a dependency or dependent will be passed on to `defer` logic (and it is considered `defer` as long as it has all it's dependents deferred )
